### ISSUE TITLE
Node versioning...

### ISSUE THREAD

**Issue by agorajek** [DataJunction/dj#262]
When we update an existing node (say change the SQL of a transform that's connected up and down the dag) we should be conscious about how we handle that:
1) do we update in place ?
2) or do we just make the nodes immutable and help creating a new node from the original one?
3) do we make +1 version (and the node name stays the same) ?

Re #1 There are some node properties that we surely can allow to be updated in place.
Re #2 It should also be easy to make copies of nodes and create new ones under new names.

So the issue comes down to #3. We can offer a version strategy in which we would support various aspects of node management with version build. Let's think of it as a managed view layer:
1. If we support ONLY one active version, it would be obvious which one should be used by consumers.
2. We could help the node author to fill in all the gaps (columns inference + data backfills) so the DAG is valid, before they flip the active version.
3. Optionally we could support the API overrides, such that the consumers would have an escape hatch if the current active version of a node is not working for them.

**Comment by samredai** [comment]
Thanks for opening this @agorajek, a lot of great points for us to think through. A working system I've seen for similar user-defined config objects usually includes four components. This may not be the best system but I figured it's good to lay this out here since it shares a good amount with how DJ is already designed now.
- `ref` - an alpha-numeric name that's used by users and other objects in the system to refer to this object. This is defined by the user at creation time and can never be changed. (equivalent to a node `name` in DJ)
- `display_name` - A very easily update-able display name that has no impact to functionality. A `ref` of "cars.sales.revenue" might have a `display_name` of "Revenue from Sales (Cars)" (doesn't currently exist in DJ)
- `uuid` - an automatically generated uuid that represents the object in a specific state. Users have no reason to be aware of it and any update to the object generates a new `uuid` for the unchanging `ref`.
- `version` - This increments whenever the object is updated and is always paired with a new definition entry with a new `uuid` but the same `ref`

To your point #1 in the first list, I think we can break that down into two lists. One list includes properties that we want to update in place without producing a new version--basically anything that has no impact to functionality and allowing rollback is not super important. Display name and description are two examples. The second list includes properties that we want to allow updating in place but definitely want that to produce a new version. The most obvious one being the query. With some clever data modeling we can enable these varying levels of flexibility.

**Comment by agorajek** [comment]
Yes, a cool little list @samredai . Agree with your observations. I am not strict about it but I think for more approachable system I propose we use the terms in parents:
- ref (or name)
- display_name (or description)
- uuid (or id)
- version (no change)

Two questions:
1. Do you think a new `uuid` always goes with a new `version` or not? I am not sure how to think about these two. Perhaps if a `version` would be enough to represent a node in a specific state? 
2. Is the internal system ID yet another reference or is it what you mean by `uuid` above?

**Comment by samredai** [comment]
1. When a new version is created (making any update to the node), we will only append a new node definition without deleting the old one so that we can support rolling back. I was thinking of the uuid as a composite key for name+version and rolling back would just be pinning the "current" pointer to a specific uuid. I guess we could also just always include lookup nodes using two fields instead of one but does that then mean the uuid is useless and we should drop that field? Not sure what the best practice is here.

2. Yeah by uuid I meant the internal system ID (something we could show to the user somewhere off to the side but they'd never really have to remember it or use it)

**Comment by shangyian** [comment]
This is an interesting problem! I spent some time sketching out a solution based on the comments above. @agorajek and @samredai let me know what you think.

We'll have two tables: `node` and `node_history`

The `node` table would have information used to uniquely identify the node. These are things that we won't allow users to change from revision to revision. It will also contain a reference to the current version.
-  `id`: uuid, primary key
- `name`: str
- `type`: node type
-  `current_version`: int
- `created_at`: timestamp

The `node_history` table will contain all the fields that users can change. This is essentially an audit log.
- `node_id`: references `id` in the `node` table
- `version`: int
- `updated_at`: timestamp
- `query`: str
- `description`: str
- `mode`: `PUBLISHED` | `DRAFT`
- PRIMARY KEY (`node_id`, `version`)

Rolling back to an older node version would just mean updating `node.current_version` to reference a different `version` of the same `node_id` in the `node_history` table. And ideally we would keep track of the user who made the change in the history table as well (i.e., add `modified_by` column), but I skipped that for now as the current system doesn't track users.

**Comment by samredai** [comment]
@shangyian I really like that design! The other nice thing about your proposal is that the logic for creating a new node and updating an existing node becomes a lot more similar. They both just add an entry to the `node_history` table, the difference then being that updating will update a `current_version` attribute of an existing entry in the node table and creating will add a new entry to the `node` table with `current_version=1`.

**Comment by agorajek** [comment]
This is great @shangyian . I am wondering if we ever want to allow users to update the node type, but I think that just as the name change this action would likely define a new node. Thoughts?

**Comment by shangyian** [comment]
I've started to implement this, but there's a lot more going on here to flesh out. Perhaps worth chatting about in the sync... I've renamed `node_history` to `noderevision`, as the latter maps closer to the fact that it's a series of versions of nodes.

We'll have to add some items to what I outlined above, as beyond just `query`, we'll also want to track changes to `tables`. We also need to store the derived fields like `parents`, `children`, `missing_parents`, and `columns` for each `NodeRevision`, since those will change depending on the parsed query AST.

### Scenarios

Let's go through two scenarios:

**Source Node + Modifying Tables**

1. A user creates a new source node with tables A and B. We'll create a `Node` and a `NodeRevision`, with `NodeRevision.node_id`=`Node.id` and `version`=1. 
2. The user modifies the source node by removing table B. We create a new `NodeRevision` (version=2) without the table reference to B and point the `current_version` on `Node` to 2.
3. The older version=1 NodeRevision continues to reference table A. 

**Transform Node + Modifying Query**

1. A user creates a new transform node Y with a query that references source node X. We'll create a `Node` and a `NodeRevision`, with `NodeRevision.node_id`=`Node.id` and version=1.
2. This now causes an issue: do we create a new `NodeRevision` for source node X, given that its `children` have now changed? I suspect we don't.
4. Now suppose the user modifies the query on this transform node and we now reference source node Z. Modifying the query will trigger a new `NodeRevision` for transform Y. That new revision will have parents referencing source node Z. But should we now (a) remove the child for source node X's current revision or (b) create a new revision for X that has no children?
